apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PgNumeric.cs

```diff
@@ -0,0 +1,487 @@
+﻿// Copyright 2022 Google LLC
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     https://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Numerics;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace Google.Cloud.Spanner.V1
+{
+    /// <summary>
+    /// Representation of PostgreSQL numeric type which has max precision of 147,455 and a max scale of 16383. 
+    /// </summary>
+    public struct PgNumeric : IEquatable<PgNumeric>, IComparable<PgNumeric>, IComparable
+    {
+        /// <summary>
+        /// The maximum number of digits before decimal point.
+        /// </summary>
+        private const int MaxDigitsBeforeDecimalPoint = 131072;
+
+        /// <summary>
+        /// The maximum scale of numeric type.
+        /// </summary>
+        private const int MaxScale = 16383;
+
+        /// <summary>
+        /// The maximum precision of numeric type.
+        /// </summary>
+        private const int MaxPrecision = MaxDigitsBeforeDecimalPoint + MaxScale + 1;
+
+        /// <summary>
+        /// The string constant for not a number (NaN).
+        /// </summary>
+        private const string NotANumber = "NaN";
+
+        /// For a numeric(precision, scale) maximum and minimum values are defined as +- ((10^precision)-1)/(10^scale), so maximum possible numeric would be numeric
+        /// with max precision and zero scale, i.e. numeric(147455, 0), and its max and min values would be ((BigInteger.Pow(10, 147455)-1)/(BigInteger.Pow(10, 0)))
+        /// Read here: https://stackoverflow.com/questions/26396856/determine-postgres-numeric-max-min-values       
+        private static readonly BigInteger s_maxValueWithoutScale = BigInteger.Pow(10, MaxPrecision) - 1;
```

Comment:
Unless there's some special-casing of a scale of 0, I'd actually expect this to be `BigInteger.Pow(10, MaxDigitsBeforeDecimalPoint) - 1` - the maximum integer value has that many digits.